pkg/query/analyze_test.go

```diff
@@ -0,0 +1,541 @@
+			Path{
+				Resource: Object{ObjectType: "document", ObjectID: "doc1"},
+				Subject:  ObjectAndRelation{ObjectType: "user", ObjectID: "alice", Relation: "..."},
+			},
+			Path{
+				Resource: Object{ObjectType: "document", ObjectID: "doc2"},
+				Subject:  ObjectAndRelation{ObjectType: "user", ObjectID: "bob", Relation: "..."},
+			},
+		)
+		empty1 := NewEmptyFixedIterator()
+		empty2 := NewEmptyFixedIterator()
+		unoptimized := NewUnion(fixed, empty1, empty2)
+
+		// Execute unoptimized tree
+		analyzeUnoptimized := make(map[string]AnalyzeStats)
+		ctxUnoptimized := &Context{
+			Context:  context.Background(),
+			Executor: LocalExecutor{},
+			Reader:   ds.SnapshotReader(datastore.NoRevision),
+			Analyze:  analyzeUnoptimized,
+		}
+
+		resources := []Object{{ObjectType: "document", ObjectID: "doc1"}}
+		subject := ObjectAndRelation{ObjectType: "user", ObjectID: "alice", Relation: "..."}
+
+		pathSeq, err := ctxUnoptimized.Check(unoptimized, resources, subject)
+		require.NoError(t, err)
+		paths, err := CollectAll(pathSeq)
+		require.NoError(t, err)
+		require.Len(t, paths, 1)
+
+		// Get aggregated stats for unoptimized
+		unoptimizedStats := AggregateAnalyzeStats(analyzeUnoptimized)
+
+		// Apply optimizations
+		optimized, changed, err := ApplyOptimizations(unoptimized, StaticOptimizations)
+		require.NoError(t, err)
+		require.True(t, changed, "Should have optimized the tree")
+
+		// Execute optimized tree
+		analyzeOptimized := make(map[string]AnalyzeStats)
+		ctxOptimized := &Context{
+			Context:  context.Background(),
+			Executor: LocalExecutor{},
+			Reader:   ds.SnapshotReader(datastore.NoRevision),
+			Analyze:  analyzeOptimized,
+		}
+
+		pathSeq, err = ctxOptimized.Check(optimized, resources, subject)
+		require.NoError(t, err)
+		paths, err = CollectAll(pathSeq)
+		require.NoError(t, err)
+		require.Len(t, paths, 1)
+
+		// Get aggregated stats for optimized
+		optimizedStats := AggregateAnalyzeStats(analyzeOptimized)
+
+		// Verify optimization improvements
+		require.Less(t, optimizedStats.CheckCalls, unoptimizedStats.CheckCalls,
+			"Optimized tree should have fewer Check calls (unoptimized: %d, optimized: %d)",
+			unoptimizedStats.CheckCalls, optimizedStats.CheckCalls)
+
+		t.Logf("Unoptimized tree had %d Check calls, optimized tree has %d Check calls (%.1f%% reduction)",
+			unoptimizedStats.CheckCalls, optimizedStats.CheckCalls,
+			100.0*float64(unoptimizedStats.CheckCalls-optimizedStats.CheckCalls)/float64(unoptimizedStats.CheckCalls))
+
+		// Verify that timing was tracked
+		require.Greater(t, unoptimizedStats.CheckTime, time.Duration(0), "Unoptimized tree should have non-zero Check time")
+		require.Greater(t, optimizedStats.CheckTime, time.Duration(0), "Optimized tree should have non-zero Check time")
+
+		t.Logf("Timing - Unoptimized: %v, Optimized: %v", unoptimizedStats.CheckTime, optimizedStats.CheckTime)
+	})
+
+	t.Run("collapsing singleton unions reduces overhead", func(t *testing.T) {
```

Comment:
I like this test, but is it more about the optimizer or the analyzer? Or does that matter too much?